### ISSUE TITLE
[SPEC] Rework `DKGWorker` to be more share-able across different async tasks

### ISSUE THREAD

**Issue by shekohex** [webb-tools/dkg-substrate#375]
## Overview 

In the context of [`DKGWorker::run`](https://github.com/webb-tools/dkg-substrate/blob/bbc5a9071d7962f59200f76d090a185d31b84b39/dkg-gadget/src/worker.rs#L1313) method we _start_ 4 Tasks to run and watch over these results, but right now we are using [`tokio::select!`](https://docs.rs/tokio/1.21.0/tokio/macro.select.html) macro, and as per its docs it outlines the following [**"running all async expressions on the current task, the expressions are able to run concurrently but not in parallel."**](https://docs.rs/tokio/1.21.0/tokio/macro.select.html#runtime-characteristics) and for us this means for the current flow for example if `finality_notifications` future/stream finished first and at the same time we have a `dkg_msg` from the `gossip_engine` stream, the select macro could pick the block notification and completely ignoring "discarding" the message from the dkg stream resulting in a completely loss of the information here.

## Research

We should instead use `tokio::spawn` for each of these tasks and run them independently from each others in parallel, this from also the above docs it mentions the following: "If parallelism is required, spawn each async expression using [tokio::spawn](https://docs.rs/tokio/1.21.0/tokio/task/fn.spawn.html) and pass the join handle to select!"

## Examples 

Here is a good example of what I do reassemble in my head of a good solution for the current issue:

```rust
pub(crate) async fn run(mut self) {
	// ... snip ...
	let finality_notifications_task = async move {
		// ...
	};
	let import_notifications_task = async move {
		// ...
	};
	let misbehaviour_task = async move {
		// ...
	};
	let error_handler_task = async move {
		// ...
	};
	let dkg_messages_task = async move {
		// ...
	};

	let handles = vec![
		tokio::spawn(finality_notifications_task),
		tokio::spawn(import_notifications_task),
		tokio::spawn(misbehaviour_task),
		tokio::spwan(error_handler_task),
		tokio::spawn(dkg_messages_task),
	];
	let _ = futures::future::select_all(handles).await;

	// ... snip ...
}

```

## Questions/Issues

- [x] Current Implementation of the `DKGWorker` is not shareable across different tasks, e.g. most functions/methods takes `&mut self` which is not feasible to share with a mutable state, we shall use async locks or other things to make it possible.
- [x] decouple the worker into smaller states when possible, so that it could be easily shareable between different tasks.

**Comment by drewstone** [comment]
How similar is this to spawning sub-services and what happens if these threads die? Will they be re-spawned / do  we have a way to handle that?

**Comment by shekohex** [comment]
> what happens if these threads die? Will they be re-spawned / do we have a way to handle that?

This a critical issue, and no, the do not get re-spawned. If we reached this state, this a fetal error and it is okay for the node to stop, the supervisor of the node should restart the whole process again.

**Comment by shekohex** [comment]
Closed by #381
